plutus-metatheory/src/RawU.lagda.md

```diff
@@ -0,0 +1,308 @@
+
+This module provides the interface between the Haskell code and the Agda code.
+It replicates the Haskell representation of Raw. In particular, for term constants,
+we need the following extension to be able to replicate the Haskell representation
+of Universes.
+
+```
+{-# OPTIONS --type-in-type #-}
+
+module RawU where
+```
+
+## Imports
+
+```
+open import Function using (_∘_)
+open import Data.Nat using (ℕ)
+open import Data.Bool using (true;false)
+open import Data.Integer using (ℤ)
+open import Data.String using (String)
+open import Data.Bool using (Bool)
+open import Relation.Binary using (DecidableEquality)
+open import Relation.Binary.PropositionalEquality using (_≡_;refl;sym;cong;cong₂)
+open import Relation.Nullary using (yes;no;¬_)
+open import Data.Unit using (⊤;tt)
+open import Data.Product using (Σ;proj₁;proj₂) renaming (_,_ to _,,_)
+
+open import Utils using (ByteString;DATA;List;_×_;_,_)
+open import Builtin using (Builtin;equals)
+open Builtin.Builtin
+
+{-# FOREIGN GHC {-# LANGUAGE GADTs #-} #-}
+{-# FOREIGN GHC import PlutusCore #-}
+{-# FOREIGN GHC import Raw #-}
+```
+
+## A (Haskell) Universe for types
+
+The following tags use type-in-type, and map more directly to the Haskell representation
+of the default universe.
+
+Tags are indexed by the real type they represent.
+The `Esc` datatype is used in the Haskell implementation to "escape" any kind into Type.
+For constants, we only care about kind *, but we need it to match the Haskell implementation.
+```
+data Esc (a : Set) : Set where
+{-# INJECTIVE Esc #-}
+{-# COMPILE GHC Esc = data Esc () #-}
+
+data Tag : Set → Set where
+  integer    : Tag (Esc ℤ)
+  bytestring : Tag (Esc ByteString)
+  string     : Tag (Esc String)
+  bool       : Tag (Esc Bool)
+  unit       : Tag (Esc ⊤)
+  pdata      : Tag (Esc DATA)
+  pair       : ∀{A B} → Tag (Esc A) → Tag (Esc B) → Tag (Esc (A × B))
+  list       : ∀{A} → Tag (Esc A) → Tag (Esc (List A))
```

Comment:
Have you tried the version that I shared via DMs? This one:

```agda
open import Data.Bool.Base
open import Data.List.Base

data TyTag : Set where
  TagBool : TyTag
  TagList : TyTag -> TyTag

⟦_⟧ : TyTag -> Set
⟦ TagBool   ⟧ = Bool
⟦ TagList a ⟧ = List ⟦ a ⟧

data STyTag : Set -> Set where
  STagBool : STyTag Bool
  STagList : ∀ {a} -> let A = ⟦ a ⟧ in STyTag A -> STyTag (List A)

data TermCon : Set where
  termCon : ∀ {a} -> let A = ⟦ a ⟧ in STyTag A -> A -> TermCon
```

Not that I have any reason to believe it would be better, just being curious.

Reply:
Yes, that was the idea that I had when I told, but I don't know how to make it work since then the compiler requires that you match `⟦_⟧` to a Haskell datatype.

Reply:
But it's an implementation detail of `STyTag`? For all the compiler knows, there's no `⟦_⟧` and `STyTag Bool` maps to `DefaultUni Bool`. Except those nasty `Esc`s... Well, I guess we could define `⟦_⟧` over some kind of `Esc` by making is a constructor rather than data type.

But perhaps not worth spending time on, given that it's just to comply with the annoying interface of the implementation.

Reply:
> But it's an implementation detail of STyTag? For all the compiler knows, there's no ⟦_⟧

It's not an implementation detail of `StyTag` if it appears in the definition of `TermCon`.

Reply:
> It's not an implementation detail of `StyTag` if it appears in the definition of `TermCon`.

I should've said "it's an implementation detail of `STyTag`, as well as `TermCon`". What I meant is that it's not exposed outside of the data type and so it's probably possible to implement suitable pattern synonyms as a bridge between Agda and Haskell.

I could try that, if you're interested. I'll probably fail spectacularly though!

Reply:
You could try if you want, but I'm not unhappy with the current solution. I'm afraid a more direct solution is going to be more fragile to refactorings. Having two separate representations (A Haskell and an Agda representation) which can be converted to each other could be beneficial to isolate one from changes in the other.